### PR TITLE
[#234] Add concurrent primitives working with State

### LINKED ISSUE #234: Add concurrent primitives using State
I want to have `modifyTVarS :: TVar s -> StateT s STM a -> STM a` because it is often more convenient than `modifyTVar`.

We had a similar function in serokell-utils, but it is too large to be depended on for such a minor purpose. I think having this thing in universum is justified as we already export many methods related to concurrency.

A similar method is worth having for `modifyMVar`.

### PR DESCRIPTION
**Pull request by Martoon-00**
Problem: existing `modifyTVar'` is not pretty convenient. You need to
e.g. remember the order of arguments, and if several transformations
are required, composing them neatly may be difficult.

Solution: provide methods working with `State`, not a lambda. This way
one can write down modifications in do-block which is convenient.

## Related issues(s)

Fixed #234 by adding the respective methods. 

## ✓ Checklist for your Pull Request

Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

- [x] I made sure my PR addresses a single concern, or multiple concerns which
      are inextricably linked. Otherwise I should open multiple PR's.
- [x] If your PR fixes/relates to an open issue then the description should
      reference this issue. See also [auto linking on
      github](https://help.github.com/articles/autolinked-references-and-urls/).

#### Related changes (conditional)

- Tests

  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation

  I checked whether I should update the docs and did so if necessary:

  - [x] [README](/README.md)
  - [x] Haddock

- Record your changes

  - [x] I added an entry to the [changelog](/CHANGES.md) if my changes are visible to the users
        and
  - [x] provided a migration guide for breaking changes if possible

#### Stylistic guide (mandatory)

- [x] My commit history is clean (only contains changes relating to my
      issue/pull request and no reverted-my-earlier-commit changes) and commit
      messages start with identifiers of related issues in square brackets.

  **Example:** `[#42] Short commit description`

  If necessary both of these can be achieved even after the commits have been
  made/pushed using [rebase and squash](https://git-scm.com/docs/git-rebase).

### CODE REVIEW

File: src/Universum/Lifted/Concurrent.hs [hunk 1]
```diff
@@ -108,3 +115,26 @@ newTVarIO = liftIO . STM.newTVarIO
...
+updateMVarS' :: MonadIO m => MVar s -> StateT s IO a -> m a
+updateMVarS' var (StateT f) =
+  liftIO . CCM.modifyMVar var $ \s -> do
+    (!s', a) <- f s
+    return (a, s')
+{-# INLINE updateMVarS' #-}
+
+-- | Like 'modifyTVar'', but modification is specified as a 'State' monad.
+updateTVar' :: TVar s -> StateT s STM a -> STM a
+updateTVar' var (StateT f) = do
```
Comment (gromakovsky) [inline src/Universum/Lifted/Concurrent.hs:135]: Is naming intentionally inconsistent between `updateTVar'` and `updateMVarS'`?

Comment (Martoon-00) [reply src/Universum/Lifted/Concurrent.hs:135]: Meh no, forgot to remove `S` suffix.

Comment (dcastro) [reply src/Universum/Lifted/Concurrent.hs:135]: pinging @Martoon-00  ^^

Comment (Martoon-00) [reply src/Universum/Lifted/Concurrent.hs:135]: Forgot to push the changes :sweat_smile:

**Review by dcastro — APPROVED** [review]

**Review by gromakovsky — APPROVED** [review]
LGTM now. CI started failing in 2 places:
1. As we can see [here](https://hackage.haskell.org/package/tasty-hedgehog-1.2.0.0/changelog), `testProperty` was deprecated. [Apprently](https://github.com/qfpl/tasty-hedgehog/pull/42) we should use `testPropertyNamed` now. If we think that PVP is good, I propose to constrain `tasty-hedgehog` version.
2. `xrefcheck` fails because `universum` is no longer in Nightly. Good to know, I'll check it.

**Comment by gromakovsky** [comment]
> `xrefcheck` fails because `universum` is no longer in Nightly. Good to know, I'll check it.

Started working on it here: https://github.com/serokell/universum/pull/252
But it will probably take a while and involve a breaking change, so feel free to ignore that link for now and merge this PR. If you do it soon, let's set version to 1.7.3 and release it?

**Comment by Martoon-00** [comment]
Thanks for looking into this!

> `xrefcheck` fails because universum is no longer in Nightly. Good to know

Hmm, that was indeed an interesting way to check that we are not dropped from LTS.

> If you do it soon, let's set version to 1.7.3 and release it?

Yep. Should I bump the version in this PR?

**Comment by gromakovsky** [comment]
> Yep. Should I bump the version in this PR?

Yes, I propose to do it here and release the new version. I can also make you a maintainer on Hackage and you'll be able to upload it yourself. Actually I'll just go ahead and do it even if you don't want :peka:

**Comment by Martoon-00** [comment]
Uploaded ) 

Looks like it.

**Pull request merged**